File: src/sw/binomial_parallel.ml
open Printf

let init_vals t v rf n =
  let h = t /. (float_of_int n) in
  let xd = (rf -. 0.5 *. (v ** 2.)) *. h in
  let xv = v *. (sqrt h) in
  let u = exp ( xd +. xv ) in
  let d = exp ( xd -. xv ) in
  let drift = exp (rf *. h) in
  let q = (drift -. d) /. (u -. d) in
  q, u, d, drift

let c_stkval n s u d j =
  s *. (u ** (n-.j)) *. (d ** j)

let gen_initial_optvals n s u d k cp =
  for j = n downto 0 do
    let stkval = c_stkval (float n) s u d (float j) in
    output_value stdout (max 0. (cp *. (stkval -. k)))
  done

let eqn q drift a b =
  ((q *. a) +. (1.0 -. q) *. b) /. drift

let apply_column v acc pos chunk q drift = 
  let v' = ref acc.(0) in
  acc.(0) <- v;
  let maxcol = min chunk pos in
  for idx = 1 to maxcol do
    let nv' = eqn q drift acc.(idx-1) !v' in
    v' := acc.(idx);
    acc.(idx) <- nv';
  done;
  if maxcol = chunk then output_value stdout acc.(maxcol)

let process_rows rowstart rowto q drift =
  output_value stdout rowto;
  let chunk = rowstart - rowto in
  let acc = Array.create (chunk+1) 0.0 in
  for pos = 0 to rowstart do
    let v = input_value stdin in
    apply_column v acc pos chunk q drift;
  done

let _ = 
  let float_arg x = float_of_string (Sys.argv.(x)) in
  let int_arg x = int_of_string (Sys.argv.(x)) in
  let [s;k;t;v;rf;cp] = List.map float_arg [1;2;3;4;5;6] in
  let [n;chunk;start] = List.map int_arg [7;8;9] in
  let q,u,d,drift = init_vals t v rf n in
  (match start with 
  |1 -> 
      output_value stdout n;
      gen_initial_optvals n s u d k cp
  |_ ->
      let rowstart = input_value stdin in
      if rowstart = 0 then
        printf "%.9f\n%!" (input_value stdin)
      else 
        let rowto = max 0 (rowstart - chunk) in
        process_rows rowstart rowto q drift
  ); flush stdout; flush stderr

File: src/sw/binomial-parallel.c
// (c) Anil Madhavapeddy <anil@recoil.org>
// In the public domain
// Parallel binomial options parser for use with Skywriting

#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <err.h>

static double u,d,drift,q;
static double s,k,t,v,rf,cp;
static int n, chunk, start;

// function nabbed from openssh, copyright openbsd @ /usr/src/usr.bin/ssh/atomicio.c
size_t
atomicio(f, fd, _s, n)
  ssize_t (*f) (int, void *, size_t);
  int fd;
  void *_s;
  size_t n;
{
  char *s = _s;
  size_t pos = 0;
  ssize_t res;

  while (n > pos) {
    res = (f) (fd, s + pos, n - pos);
    switch (res) {
    case -1:
      if (errno == EINTR || errno == EAGAIN)
        continue;
      return 0;
    case 0:
      errno = EPIPE;
      return pos;
    default:
      pos += (size_t)res;
    }
  }
  return pos;
}

#define SAFE_IO(fn,fd,v) \
  do { \
     ssize_t r = atomicio(fn,fd, (void *)(&(v)), sizeof((v))); \
     if (r != (sizeof((v)))) err(1,#fn); \
  } while(0)
#define vwrite (ssize_t (*)(int, void *, size_t))write
#define INPUT(v) SAFE_IO(read,STDIN_FILENO,v)
#define OUTPUT(v) SAFE_IO(vwrite,STDOUT_FILENO,v)

void
init_vals(void)
{
  double h = t / (double)n;
  double xd = (rf - 0.5 * pow(v,2)) * h;
  double xv = v * sqrt(h);
  u = exp(xd + xv);
  d = exp(xd - xv);
  drift = exp(rf * h);
  q = (drift - d) / (u - d);
}  

void
gen_initial_optvals(void)
{
  double stkval;
  for (int j=n; j>-1; j--) {
    stkval = s * pow(u,(n-j)) * pow(d,j);
    stkval = fmax(0, (cp * (stkval-k)));
    OUTPUT(stkval);
  }
}

void process_rows(int rowstart, int rowto)
{
  OUTPUT(rowto);
  int chunk = rowstart - rowto;
  double *acc = malloc((chunk+1)*sizeof(double));
  if (!acc) err(1, "malloc");
  for (int pos=0; pos<=rowstart; pos++) {
    double v,v2;
    INPUT(v);
    v2=acc[0];
    acc[0]=v;
    int maxcol = chunk < pos ? chunk : pos;
    for (int idx=1; idx<=maxcol; idx++) {
      double nv = ((q * acc[idx-1]) + (1-q) * v2) / drift;
      v2 = acc[idx];
      acc[idx] = nv;
    }
    if (maxcol == chunk)
      OUTPUT(acc[maxcol]);
  }
}

int 
main(int argc, char **argv)
{
  s = strtod(argv[1], NULL); 
  k = strtod(argv[2], NULL);
  t = strtod(argv[3], NULL);
  v = strtod(argv[4], NULL);
  rf = strtod(argv[5], NULL);
  cp = strtod(argv[6], NULL);
  n = (int)strtod(argv[7], NULL);
  chunk = (int)strtod(argv[8], NULL);
  start = (int)strtod(argv[9], NULL);
  init_vals();
  if (start == 1) {
    OUTPUT(n);
    gen_initial_optvals();
  } else {
    int rowstart=0;
    INPUT(rowstart);
    if (rowstart==0) {
      double v;
      INPUT(v);
      printf("%.9f\n", v);
    } else {
      int rowto = (rowstart - chunk) > 0 ? (rowstart - chunk) : 0;
      process_rows(rowstart, rowto);
    }
  }
  return 0;
}
